File: src/listas/basica/vPython/List_formal.py
from Node_formal import Node

class List:
    def __init__(self):
        self._first = None

    def size(self):
        count = 0
        iterator = self._first
        while iterator is not None:
            count += 1
            iterator = iterator.next
        return count

    def is_empty(self):
        return self.size() == 0

    def insert_end(self, value):
        new_node = Node(value)
        if not self._first:
            self._first = new_node
        else:
            iterator = self._first
            while iterator.next is not None:
                iterator = iterator.next
            iterator.next = new_node

    def insert_front(self, value):
        new_node = Node(value)
        if not self._first:
            self._first = new_node
        else:
            new_node.next = self._first
            self._first = new_node

    def delete(self):
        if self._first:
            if self._first.next is None:
                self._first = None
            else:
                iterator = self._first
                while iterator.next.next is not None:
                    iterator = iterator.next
                iterator.next = None

    def list_all(self):
        items = []
        iterator = self._first
        while iterator is not None:
            items.append(iterator.value)
            iterator = iterator.next
        return items

File: production/eda_23_24/listas/basica/vJavascript/List_formal.js
const Node = require('./Node_formal');

class List {
    constructor() {
        this.first = null;
    }

    size() {
        let count = 0;
        let iterator = this.first;
        while (iterator !== null) {
            count++;
            iterator = iterator.next;
        }
        return count;
    }

    isEmpty() {
        return this.size() === 0;
    }

    insertEnd(value) {
        const newNode = new Node(value);
        if (this.first === null) {
            this.first = newNode;
        } else {
            let iterator = this.first;
            while (iterator.next !== null) {
                iterator = iterator.next;
            }
            iterator.next = newNode;
        }
    }

    insertFront(value) {
        const newNode = new Node(value);
        newNode.next = this.first;
        this.first = newNode;
    }

    delete() {
        if (this.first !== null) {
            let iterator = this.first;
            let previous = null;
            while (iterator.next !== null) {
                previous = iterator;
                iterator = iterator.next;
            }
            if (previous) {
                previous.next = null;
            } else {
                this.first = null;
            }
        }
    }

    listAll() {
        const list = [];
        let iterator = this.first;
        while (iterator !== null) {
            list.push(iterator.value);
            iterator = iterator.next;
        }
        return list;
    }
}

module.exports = List;
